### PR TITLE
Force to use the stage 2 compiler  when config.download_rustc is set

### LINKED ISSUE #108768: `x test --stage 2 std` uses the wrong stage when download-rustc is enabled
cc https://github.com/rust-lang/rust/issues/81930

I tried this code:

```rust
./configure --set rust.download-rustc
x test --stage 2 library/core
```

I expected to see this happen: Bootstrap uses the stage 2 compiler to run the tests.

Instead, this happened: Bootstrap uses the stage 1 compiler:
```
Testing {core} stage1 (aarch64-unknown-linux-gnu -> aarch64-unknown-linux-gnu)
   Compiling core v0.0.0 (/home/gh-jyn514/rust/library/core)
error[E0773]: attempted to define built-in macro more than once
    --> /home/gh-jyn514/rust/library/core/src/macros/mod.rs:1310:5
     |
1310 |     macro_rules! cfg {
     |     ^^^^^^^^^^^^^^^^
     |
note: previously defined here
    --> /home/gh-jyn514/rust/build/aarch64-unknown-linux-gnu/ci-rustc-sysroot/lib/rustlib/src/rust/library/core/src/macros/mod.rs:13
10:5
     |
1310 |     macro_rules! cfg {
     |     ^^^^^^^^^^^^^^^^

error[E0152]: duplicate lang item in crate `core` (which `corebenches` depends on): `sized`.
  |
  = note: the lang item is first defined in crate `core` (which `std` depends on)
  = note: first definition in `core` loaded from /home/gh-jyn514/rust/build/aarch64-unknown-linux-gnu/ci-rustc-sysroot/lib/rustlib/a
arch64-unknown-linux-gnu/lib/libcore-d20cd64dc93f727e.rlib
  = note: second definition in `core` loaded from /home/gh-jyn514/rust/build/aarch64-unknown-linux-gnu/stage1-std/aarch64-unknown-li
nux-gnu/release/deps/libcore-8c941efb0bc89716.rlib
```

I think the right fix is to unconditionally use the stage 2 compiler in `compiler_for` when `config.download_rustc` is set:
https://github.com/rust-lang/rust/blob/2f163554866e1b4b348811e9df6a4a753beb0abf/src/bootstrap/test.rs#L2103
https://github.com/rust-lang/rust/blob/a5d27316fe49feb36c332f16dd82da959de664fb/src/bootstrap/lib.rs#L1206-L1210

### Meta

HEAD is https://github.com/rust-lang/rust/commit/35636f9459720513ca4ed19373a4a7c9e0ea3c46

### PR DESCRIPTION
**Pull request by Kohei316**
Fixes #108768

### CODE REVIEW

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -1209,6 +1209,10 @@ impl Build {
             && (self.hosts.iter().any(|h| *h == target) || target == self.build)
     }
```
Comment (albertlarsan68) [inline src/bootstrap/lib.rs:1211]:
```suggestion

    /// Whether to force the use of stage 2 compiler in any case
```

I think starting to add doc-comments to all new things will help in the future.

**Review by onur-ozkan — APPROVED** [review]
thank you for the fix, please consider suggestion from @albertlarsan68, I will queue the PR after that

**Comment by ozkanonur** [comment]
can you squash your commits into single commit?

File: src/bootstrap/lib.rs [hunk 2]
```diff
@@ -1209,6 +1209,15 @@ impl Build {
...
 
+    /// Tests whether the `compiler` compiling for `target` should be forced to
+    /// use a stage2 compiler instead.
+    ///
+    /// If we use a downloaded, pre-compiled version of rustc it should be forced to
+    /// use a stage2 compiler.
```
Comment (ozkanonur) [inline src/bootstrap/lib.rs:1216]:
```suggestion
    /// Checks whether the `compiler` compiling for `target` should be forced to
    /// use a stage2 compiler instead.
    ///
    /// When we download the pre-compiled version of rustc it should be forced to
    /// use a stage2 compiler.
```

**Review by onur-ozkan — APPROVED** [review]
Waiting for CI to be green

**Comment by jsha** [comment]
This PR may have caused a failure running tests for some users: https://github.com/rust-lang/rust/issues/109286

```
Uplifting library (stage2 -> stage0)
Error: Unable to find the stamp file ...rust/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/release/.libstd.stamp, did you try to keep a nonexistent build stage?
```

**Pull request merged**